src/libstd/io/buffered.rs

```diff
@@ -451,7 +453,14 @@ impl<W: Write> BufWriter<W> {
     pub fn into_inner(mut self) -> Result<W, IntoInnerError<BufWriter<W>>> {
         match self.flush_buf() {
             Err(e) => Err(IntoInnerError(self, e)),
-            Ok(()) => Ok(self.inner.take().unwrap())
+            Ok(()) => unsafe {
+                // use `ptr::read` to extract the inner writer and circumvent
+                // our `Drop` implementation
+                mem::replace(&mut self.buf, Vec::new());
```

Comment:
To be honest, I find using `mem::replace` here pretty confusing - in my opinion `ptr::read`ing the buffer (maybe with a comment why) would convey the intent much better (dropping the buffer), especially since we rely on the fact that `Vec::new()` never allocates (which *is* document, but still).